### ISSUE TITLE
File/stdin cleanup not happening when using shmem fuzzing

### ISSUE THREAD

**Issue by k3an3** [AFLplusplus/AFLplusplus#1616]
**Describe the bug**
EDIT: My apologies if this doesn't make any sense. Basically, my fuzzed program's stdout/stderr is getting written to the `.cur_input` file and filling up my ramdisk very quickly. My assumption is that it was due to the dangling file descriptor from when AFL opens that file during `setup_stdio_file()`. Ultimately, the fuzzed program is causing the problem, but it seems that AFL doesn't need to have this file handle open either.

I noticed that I was getting a file at for each fuzzer instance at `out_dir/fuzzer/.cur_input` that kept growing indefinitely (until my ramdisk fills up), even though I am using shared memory fuzzing and wouldn't expect to see a file written at all. I haven't seen this happen before, so maybe there is a problem with my specific harness.

That said, I do think there is a bug in AFLplusplus too:

If no `-f` argument is specified nor `@@`, a file (.cur_input) is created for fuzzing, and the file descriptor is saved:
https://github.com/AFLplusplus/AFLplusplus/blob/75fb918a37ad29b93bd6c220460404dae156fa8c/src/afl-fuzz.c#L1940
https://github.com/AFLplusplus/AFLplusplus/blob/75fb918a37ad29b93bd6c220460404dae156fa8c/src/afl-fuzz-init.c#L2112-L2128

Later, shmem is enabled:
https://github.com/AFLplusplus/AFLplusplus/blob/75fb918a37ad29b93bd6c220460404dae156fa8c/src/afl-fuzz.c#L1978
<strike>Several things are set on the `afl` pointer:
https://github.com/AFLplusplus/AFLplusplus/blob/75fb918a37ad29b93bd6c220460404dae156fa8c/src/afl-fuzz-init.c#L2580-L2582
</strike>

Lastly, some cleanup is supposed to happen if shmem is enabled, introduced in #1541 :
https://github.com/AFLplusplus/AFLplusplus/blob/75fb918a37ad29b93bd6c220460404dae156fa8c/src/afl-fuzz.c#L2155-L2158

Two issues:
~1. I don't think this conditional is ever true. `afl->fsrv.use_shmem_fuzz` has never been set (and is not set until `afl-forkserver.c:871`), though the similarly-named `afl->fsrv.support_shmem_fuzz` was set in `setup_testcase_shmem()`.~
2. `afl->fsrv.out_fd` that was opened in `setup_stdio_file()` is left open

**Proposed fixes**
<strike>1. afl-fuzz.cc:2155 should be changed to this perhaps:
```c
  if (afl->fsrv.out_file && afl->fsrv.support_shmem_fuzz) {
```
Or, maybe `afl->fsrv.use_shmem_fuzz` should get set to 1 by `setup_testcase_shmem()`.</strike>

2. Maybe don't set up stdio fuzzing unless we are sure we are going to use it. Or close the `afl->fsrv.out_fd` if shmem is being used.

**Comment by k3an3** [comment]
Oops, looks like I missed the call to `afl_fsrv_start()` which does indeed set `afl->fsrv.use_shmem_fuzz = 1` before that conditional. However, there is still the fd that is leaked if stdio fuzzing is not used.

**Comment by vanhauser-thc** [comment]
I made a little fix: 7abbc8d7401e4a358986a5ff5d1157f44761e6a7
looks good now?
